File: core/src/main/java/nl/t64/game/rpg/GameData.java
package nl.t64.game.rpg;

import lombok.Getter;
import nl.t64.game.rpg.components.inventory.GlobalContainer;
import nl.t64.game.rpg.components.inventory.InventoryDatabase;
import nl.t64.game.rpg.components.inventory.InventoryItem;
import nl.t64.game.rpg.components.party.HeroContainer;
import nl.t64.game.rpg.components.party.HeroItem;
import nl.t64.game.rpg.components.party.PartyContainer;
import nl.t64.game.rpg.constants.Constant;
import nl.t64.game.rpg.profile.ProfileManager;
import nl.t64.game.rpg.profile.ProfileObserver;


@Getter
public class GameData implements ProfileObserver {

    private HeroContainer heroes;
    private PartyContainer party;
    private GlobalContainer inventory;

    @Override
    public void onNotifyCreate(ProfileManager profileManager) {
        heroes = new HeroContainer();
        party = new PartyContainer();
        inventory = new GlobalContainer();
        addFirstHeroToParty();
        addFirstItemsToInventory();
        onNotifySave(profileManager);
    }

    @Override
    public void onNotifySave(ProfileManager profileManager) {
        profileManager.setProperty("heroes", heroes);
        profileManager.setProperty("party", party);
        profileManager.setProperty("inventory", inventory);
    }

    @Override
    public void onNotifyLoad(ProfileManager profileManager) {
        heroes = profileManager.getProperty("heroes", HeroContainer.class);
        party = profileManager.getProperty("party", PartyContainer.class);
        inventory = profileManager.getProperty("inventory", GlobalContainer.class);
    }

    private void addFirstHeroToParty() {
        HeroItem hero = heroes.getHero(Constant.PLAYER_ID);
        heroes.removeHero(Constant.PLAYER_ID);
        party.addHero(hero);
    }

    private void addFirstItemsToInventory() {
        InventoryItem gold = InventoryDatabase.getInstance().getInventoryItem("gold");
        InventoryItem mace = InventoryDatabase.getInstance().getInventoryItem("basic_mace");
        inventory.forceSetItemAt(0, gold);
        inventory.forceSetItemAt(1, mace);
    }

}

File: core/src/nl/t64/game/rpg/GameData.kt
package nl.t64.game.rpg

import nl.t64.game.rpg.components.battle.BattleContainer
import nl.t64.game.rpg.components.conversation.ConversationContainer
import nl.t64.game.rpg.components.conversation.PhraseIdContainer
import nl.t64.game.rpg.components.cutscene.CutsceneContainer
import nl.t64.game.rpg.components.door.DoorContainer
import nl.t64.game.rpg.components.event.EventContainer
import nl.t64.game.rpg.components.loot.LootContainer
import nl.t64.game.rpg.components.loot.SpoilsContainer
import nl.t64.game.rpg.components.party.HeroContainer
import nl.t64.game.rpg.components.party.PartyContainer
import nl.t64.game.rpg.components.party.inventory.InventoryContainer
import nl.t64.game.rpg.components.party.inventory.InventoryDatabase.createInventoryItem
import nl.t64.game.rpg.components.quest.QuestContainer
import nl.t64.game.rpg.constants.Constant
import nl.t64.game.rpg.subjects.ProfileObserver


class GameData : ProfileObserver {

    lateinit var heroes: HeroContainer
    lateinit var party: PartyContainer
    lateinit var inventory: InventoryContainer
    lateinit var battles: BattleContainer
    lateinit var conversations: ConversationContainer
    lateinit var quests: QuestContainer
    lateinit var events: EventContainer
    lateinit var loot: LootContainer
    lateinit var spoils: SpoilsContainer
    lateinit var doors: DoorContainer
    lateinit var cutscenes: CutsceneContainer
    var isTooltipEnabled = false
    var isComparingEnabled = false

    override fun onNotifyCreateProfile(profileManager: ProfileManager) {
        heroes = HeroContainer()
        party = PartyContainer()
        inventory = InventoryContainer()
        battles = BattleContainer()
        conversations = ConversationContainer()
        quests = QuestContainer()
        events = EventContainer()
        loot = LootContainer()
        spoils = SpoilsContainer()
        doors = DoorContainer()
        cutscenes = CutsceneContainer()
        isTooltipEnabled = true
        isComparingEnabled = true
        addFirstHeroToParty()
        addFirstItemsToInventory()
        onNotifySaveProfile(profileManager)
    }

    override fun onNotifySaveProfile(profileManager: ProfileManager) {
        profileManager.setProperty("heroes", heroes)
        profileManager.setProperty("party", party)
        profileManager.setProperty("inventory", inventory)
        profileManager.setProperty("battles", battles)
        profileManager.setProperty("conversations", conversations.createPhraseIdContainer())
        profileManager.setProperty("quests", quests)
        profileManager.setProperty("events", events)
        profileManager.setProperty("loot", loot)
        profileManager.setProperty("spoils", spoils)
        profileManager.setProperty("doors", doors)
        profileManager.setProperty("cutscenes", cutscenes)
        profileManager.setProperty("isTooltipEnabled", isTooltipEnabled)
        profileManager.setProperty("isComparingEnabled", isComparingEnabled)
    }

    override fun onNotifyLoadProfile(profileManager: ProfileManager) {
        heroes = profileManager.getProperty("heroes")
        party = profileManager.getProperty("party")
        inventory = profileManager.getProperty("inventory")
        battles = profileManager.getProperty("battles")
        conversations = ConversationContainer().apply {
            val container: PhraseIdContainer = profileManager.getProperty("conversations")
            this.setCurrentPhraseIds(container)
        }
        quests = profileManager.getProperty("quests")
        events = profileManager.getProperty("events")
        loot = profileManager.getProperty("loot")
        spoils = profileManager.getProperty("spoils")
        doors = profileManager.getProperty("doors")
        cutscenes = profileManager.getProperty("cutscenes")
        isTooltipEnabled = profileManager.getProperty("isTooltipEnabled")
        isComparingEnabled = profileManager.getProperty("isComparingEnabled")
    }

    private fun addFirstHeroToParty() {
        val hero = heroes.getCertainHero(Constant.PLAYER_ID)
        heroes.removeHero(Constant.PLAYER_ID)
        party.addHero(hero)
    }

    private fun addFirstItemsToInventory() {
        val mace = createInventoryItem("basic_mace")
        val gold = createInventoryItem("gold")
        inventory.autoSetItem(mace)
        inventory.autoSetItem(gold)
    }

}
